### PR TITLE
#630 Two VRRemote with same name are added when a node is regulated by several generators and a HVDC with vscs

### PR DESCRIPTION
**Pull request by barondim**
**Checklist before requesting a review**

*use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes*

- [x] unit tests and non regression tests were added (update of inputs, outputs, SA, algos)
- [ ] main documentation was updated (update of input/output file)
- [x] the corresponding milestone was added in the ticket and in this PR

### CODE REVIEW

File: sources/Outputs/src/DydVRRemote.cpp [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        regulatedBusIds.insert(busToGenerator.first);
+    for (const auto& busToVSC : vscBusVSCDefinitionsMap)
+        regulatedBusIds.insert(busToVSC.first);
+
+    for (const BusId& regulatedBusId : regulatedBusIds) {
+        std::string id = constants::modelSignalNQprefix_ + regulatedBusId;
+        boost::shared_ptr<dynamicdata::BlackBoxModel> blackBoxModelVRRemote =
+                                                        helper::buildBlackBox(id, "VRRemote", basename + ".par", id);
+        dynamicModelsToConnect->addModel(blackBoxModelVRRemote);
+        dynamicModelsToConnect->addConnect(id, "vrremote_URegulated", constants::networkModelName, regulatedBusId + "_U_value");
```
Comment (rosiereflo) [inline sources/Outputs/src/DydVRRemote.cpp:36]: the connections to the generators/hvdc are missing
image

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: sources/Outputs/src/DydVRRemote.cpp (hunk 1 above)
Comment (rosiereflo) [reply sources/Outputs/src/DydVRRemote.cpp:36]: Yes, this is the point of this development

File: sources/Context.h [hunk 2]
```diff
@@ -146,7 +146,7 @@ class Context {
    * @param cbk the callback to register
    */
   void onNodeOnMainConnexComponent(ProcessNodeCallBackMainComponent &&cbk) {
-    callbacksMainConnexComponent_.push_back(std::forward<ProcessNodeCallBackMainComponent>(cbk));
```
Comment (rosiereflo) [inline sources/Context.h:149]: I'm quite sure we need to use forward here. Anyway this is not related so please revert this change.

File: sources/Outputs/include/Dyd.h [hunk 3]
```diff
@@ -70,27 +71,39 @@ class Dyd {
...
+        dydSVarC_(new DydSVarC(svarcsDefs)),
+        dydHvdc_(new DydHvdc(hvdcDefinitions)),
+        dydGenerator_(new DydGenerator(gens)),
+        dydDynModel_(new DydDynModel(models, gens, loaddefs)),
+        dydVRRemote_(new DydVRRemote(gens, hvdcDefinitions)) {}
 
     std::string basename_;                                                                    ///< basename for file
     std::string filename_;                                                                    ///< filepath for file to write
     std::shared_ptr<inputs::Node> slackNode_;                                                 ///< slack node to use
-    const algo::HVDCLineDefinitions &hvdcDefinitions_;                                        ///< list of hvdc definitions
```
Comment (rosiereflo) [inline sources/Outputs/include/Dyd.h:86]: wrong merge?

File: sources/Outputs/src/DydVRRemote.cpp [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+      dynamicModelsToConnect->addMacroConnect(connection);
+    }
+  }
+}
+
+void
+DydVRRemote::writeConnect(boost::shared_ptr<dynamicdata::DynamicModelsCollection>& dynamicModelsToConnect) {
+  const std::string vrremoteNqValue("vrremote_NQ");
+  for (const auto& keyValue : hvdcDefinitions_.hvdcLines) {
+    algo::HVDCDefinition hvdcLine = keyValue.second;
```
Comment (rosiereflo) [inline sources/Outputs/src/DydVRRemote.cpp:79]: why 2 methods to create connections?

File: sources/Outputs/src/ParVRRemote.cpp [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+      busesRegulatedBySeveralGeneratorsToFrozen[generator.regulatedBusId] = false;
+    }
+  }
+
+  // delete BusId already in busesRegulatedBySeveralGenerators_ to not generate VRRemotes twice
+  algo::HVDCLineDefinitions::BusVSCMap vscBusVSCDefinitionsMapWithoutBusDuplicates(vscBusVSCDefinitionsMap_);
+  for (const auto& keyValue : busesRegulatedBySeveralGenerators_) {
+    if (vscBusVSCDefinitionsMapWithoutBusDuplicates.find(keyValue.first) != vscBusVSCDefinitionsMapWithoutBusDuplicates.end()) {
+      algo::HVDCDefinition::BusId busId = static_cast<algo::HVDCDefinition::BusId>(keyValue.first);
+      vscBusVSCDefinitionsMapWithoutBusDuplicates.erase(busId);
```
Comment (rosiereflo) [inline sources/Outputs/src/ParVRRemote.cpp:41]: Why do we need an intermediate container? We could do the test on the fly below?

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: sources/Context.h (hunk 2 above)
Comment (barondim) [reply sources/Context.h:149]: ProcessNodeCallBackMainComponent&& isn't a universal reference. It's a rvalue reference so I thought std::move was more appropriate.

Comment (rosiereflo) [reply sources/Context.h:149]: yes it is a universal reference to a rvalue within a template so forward is required.

Comment (barondim) [reply sources/Context.h:149]: std::function<void(const std::shared_ptr<inputs::Node> &, std::shared_ptr<dfl::algo::AlgorithmsResults> &)> isn't a template, right ?

Comment (rosiereflo) [reply sources/Context.h:149]: we can discuss this later on case by case. Anyway this is not related to this development so please revert this in the context of this PR.

File: sources/Outputs/include/Dyd.h (hunk 3 above)
Comment (barondim) [reply sources/Outputs/include/Dyd.h:86]: This attribute is just never used

File: sources/Outputs/src/DydVRRemote.cpp (hunk 4 above)
Comment (barondim) [reply sources/Outputs/src/DydVRRemote.cpp:79]: Currently, the connections creation is splited in 3 methods : 

writeMacroConnector(dynamicModelsToConnect);
writeMacroConnect(dynamicModelsToConnect);
writeConnect(dynamicModelsToConnect);

but I can merge them into one big method : writeAllConnections

Comment (rosiereflo) [reply sources/Outputs/src/DydVRRemote.cpp:79]: writeMacroConnector can be keep as it creates the macros, however the two others can be merged IMHO

File: sources/Outputs/src/ParVRRemote.cpp (hunk 5 above)
Comment (barondim) [reply sources/Outputs/src/ParVRRemote.cpp:41]: We can but I thought it could be risky to alter directly vscBusVSCDefinitionsMap_ ^^'

Comment (rosiereflo) [reply sources/Outputs/src/ParVRRemote.cpp:41]: This is not my proposal. I am proposing to test if the bus was already implemented directly in the hvdc loop

Comment (barondim) [reply sources/Outputs/src/ParVRRemote.cpp]: Yes, you are right. I could be optimized.

File: sources/Outputs/include/ParVRRemote.h [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    generatorDefinitions_(generatorDefinitions),
+    busesRegulatedBySeveralGenerators_(busesRegulatedBySeveralGenerators),
+    vscBusVSCDefinitionsMap_(vscBusVSCDefinitionsMap) {}
+
+  /**
+   * @brief enrich the parameter set collection for VRRemote
+   *
+   * @param paramSetCollection parameter set collection to enrich
+   */
+  void writeVRRemotes(boost::shared_ptr<parameters::ParametersSetCollection>& paramSetCollection);
```
Comment (rosiereflo) [inline sources/Outputs/include/ParVRRemote.h:52]: Should be private IMHO

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: sources/Outputs/include/ParVRRemote.h (hunk 6 above)
Comment (barondim) [reply sources/Outputs/include/ParVRRemote.h:52]: If this method is private, we can't call it in Par::write()

File: tests/outputs/TestDyd.cpp [hunk 7]
```diff
@@ -316,3 +316,56 @@ TEST(Dyd, writeLoad) {
...
+                                                                outputPath.generic_string(),
+                                                                {},
+                                                                {},
+                                                                node,
+                                                                hvdcDefs,
+                                                                busesRegulatedBySeveralGenerators,
+                                                                manager,
+                                                                noModels,
+                                                                {}));
+
```
Comment (rosiereflo) [inline tests/outputs/TestDyd.cpp:363]: I don't see any unit test where there are both hvdc and generators regulated by a VRRemote?

Comment (barondim) [reply tests/outputs/TestDyd.cpp:363]: In this unit test :

dydWriter.def_.dydVRRemote_->hvdcDefinitions_.vscBusVSCDefinitionsMap contains BUS_1 and BUS_3

dydWriter.def_.dydVRRemote_->busesRegulatedBySeveralGenerators_ contains BUS_1 and BUS_2

The unit test checks that only one Model_Signal_NQ is created for BUS_1 even if it's both regulated by a HVDC and a generator

**Review by rosiereflo — APPROVED** [review]
